### PR TITLE
[DUOS-2157][risk=low] Add Dataset Batch GET Endpoint

### PR DESCRIPTION
**Pull request by connorlbark**
### Addresses

https://broadworkbench.atlassian.net/browse/DUOS-2157

----
Have you read [CONTRIBUTING.md](../CONTRIBUTING.md) lately? If not, do that first.

- Label PR with a Jira ticket number and include a link to the ticket
- Label PR with a security risk modifier [no, low, medium, high]
- PR describes scope of changes
- Get a minimum of one thumbs worth of review, preferably two if enough team members are available
- Get PO sign-off for all non-trivial UI or workflow changes
- Verify all tests go green
- Test this change deployed correctly and works on dev environment after deployment

### CODE REVIEW

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java [hunk 1]
```diff
@@ -673,6 +674,46 @@ public void testGetDatasetNotFound() {
...
+
+        when(datasetService.getDatasets(List.of(1,2,3,4))).thenReturn(List.of(
+                ds1,
+                ds3
+        ));
+
+        initResource();
+        Response response = resource.getDatasets(List.of(1,2,3,4));
+        assertEquals(404, response.getStatus());
+        assertEquals("Could not find datasets with ids: 2,4", ((Error)response.getEntity()).getMessage());
```
Comment (connorlbark) [inline src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java]: is there a more efficient way to check error message?

File: src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java [hunk 2]
```diff
@@ -280,6 +280,35 @@ public Response getDataset(@PathParam("datasetId") Integer datasetId){
...
     }
 
+    @GET
+    @Path("/batch")
+    @Produces("application/json")
+    @PermitAll
+    public Response getDatasets(@QueryParam("ids") List<Integer> datasetIds){
+        try {
+            List<Dataset> datasets = datasetService.getDatasets(datasetIds);
+            if (datasets.size() != datasetIds.size()) {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java]: Size is one way to do this, but can be an error in the case where a user provides the same id more than once (i.e. ` 1, 1, 2, 3` -> list of datasets will be 3). 

One thing to do would be to ensure distinctness on the provided list.

Another option might look something like this:
```
List<Integer> foundIds = datasets.stream().map(Dataset::getDataSetId).toList();
if (!foundIds.containsAll(datasetIds)) { 
  // find the differences
  List<Integer> differences = new ArrayList<>(datasetIds);
  differences.removeAll(foundIds);
  ... throw error with differences ... 
}
```

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java (hunk 1 above)
Comment (rushtong) [reply src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java]: IMHO, message testing is not very helpful and often brittle so I discourage it. But a direct answer is no, reading the entity is the only way to get the content out of the response.

File: src/main/resources/assets/paths/datasetBatch.yaml [hunk 3]
```diff
@@ -0,0 +1,27 @@
+get:
+  summary: Get Dataset by Id
```
Comment (rushtong) [inline src/main/resources/assets/paths/datasetBatch.yaml]:
```suggestion
  summary: Get Dataset by list of Ids
```

File: src/main/resources/assets/paths/datasetBatch.yaml [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+  description: Finds Dataset and DatasetProperties of the Dataset specified by id.
```
Comment (rushtong) [inline src/main/resources/assets/paths/datasetBatch.yaml:3]:
```suggestion
  description: Finds Dataset and DatasetProperties of the Dataset specified by list of ids.
```

File: src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java (hunk 2 above)
Comment (rushtong) [reply src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java]: Testing with a duplicate ID that is valid shows this response:
```
{
  "message": "Could not find datasets with ids: ",
  "code": 404
}
```

File: src/main/resources/assets/paths/datasetBatch.yaml [hunk 5]
```diff
@@ -0,0 +1,27 @@
...
+  parameters:
+    - name: ids
+      in: query
+      description: Comma separate dataset ids, e.g. "1,2,3,4,5"
+      required: true
+      schema:
+        type: array
+        items:
+          type: integer
```
Comment (rushtong) [inline src/main/resources/assets/paths/datasetBatch.yaml]: The curl call shows a different format: `/api/dataset/batch?ids=2&ids=369`

If, instead, I use a comma separated list this way, Jersey doesn't actually parse the list in the way that is described in the docs:
```
curl -X 'GET' 'https://local.broadinstitute.org:27443/api/dataset/batch?ids=2,369'
```
yields
```
{"code":404,"message":"HTTP 404 Not Found"}
```
since that endpoint doesn't exist

File: src/main/resources/assets/paths/datasetBatch.yaml [hunk 6]
```diff
@@ -0,0 +1,27 @@
...
+  tags:
+    - Dataset
+  responses:
+    200:
+      description: The Dataset with properties
```
Comment (rushtong) [inline src/main/resources/assets/paths/datasetBatch.yaml]:
```suggestion
      description: The list of Datasets with properties
```

**Review by rushtong — CHANGES_REQUESTED** [review]
Looking good! Some doc updates and an edge case to consider inline.

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java (hunk 1 above)
Comment (connorlbark) [reply src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java]: Hmm, I totally feel you which is why I was unsure of this test. However, I wanted to make sure the logic to return just the datasets not found worked properly. I considered making it a helper function, but I couldn't figure out a logical place to make that a public/testable function. So I leaned on testing the message directly instead.

**Comment by connorlbark** [comment]
@rushtong thanks for the helpful feedback !! I think I addressed it all

File: src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java [hunk 7]
```diff
@@ -280,6 +281,30 @@ public Response getDataset(@PathParam("datasetId") Integer datasetId){
...
+    @GET
+    @Path("/batch")
+    @Produces("application/json")
+    @PermitAll
+    public Response getDatasets(@QueryParam("ids") List<Integer> datasetIds){
+        try {
+            List<Dataset> datasets = datasetService.getDatasets(datasetIds);
+
+            List<Integer> foundIds = datasets.stream().map(Dataset::getDataSetId).toList();
+            if (!foundIds.containsAll(datasetIds)) {
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java:294]: So, I may have steered you poorly - I didn't actually test the code I used as an example. Turns out `List. containsAll()` [has poor performance](https://stackoverflow.com/questions/10199772/what-is-the-cost-of-containsall-in-java). The suggestion is to use a hash instead,
```
if (!new HashSet<>(foundIds).containsAll(datasetIds)) {
```

File: src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java [hunk 8]
```diff
@@ -280,6 +281,30 @@ public Response getDataset(@PathParam("datasetId") Integer datasetId){
...
+                // find the differences
+                List<Integer> differences = new ArrayList<>(datasetIds);
+                differences.removeAll(foundIds);
+                throw new NotFoundException(
+                        "Could not find datasets with ids: "
+                                + String.join(",", differences.stream().map((i) -> i.toString()).collect(Collectors.toSet())));
+
+            }
+            return Response.ok(datasets, MediaType.APPLICATION_JSON).build();
```
Comment (rushtong) [inline src/main/java/org/broadinstitute/consent/http/resources/DatasetResource.java]: One more thing, this can be simpler, `MediaType.APPLICATION_JSON` is already specified on line 286.

```
return Response.ok(datasets).build();
```
or 
```
return Response.ok().entity(datasets).build();
```
In the above cases, Jackson will coerce the objects to JSON format based on `@Produces`

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java (hunk 1 above)
Comment (rushtong) [reply src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java]: One thing you could do is test for the existence/non-existence of the IDs in the string instead of looking at complete string equality.

**Review by rushtong — COMMENTED** [review]
The updates look good, please see additional (optional) comments inline.

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java (hunk 1 above)
Comment (connorlbark) [reply src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java]: ahh good idea, just implemented I believe a decent way of doing this.

**Comment by connorlbark** [comment]
@rushtong should be ready for review again !

File: src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java [hunk 9]
```diff
@@ -695,6 +697,48 @@ public void testGetDatasets() {
...
+                ds1,
+                ds2
+        ));
+
+        initResource();
+        Response response = resource.getDatasets(List.of(1,1,2,2,3,3));
+        assertEquals(404, response.getStatus());
+        assertTrue(((Error)response.getEntity()).getMessage().contains("3"));
+        assertFalse(((Error)response.getEntity()).getMessage().contains("2"));
+        assertFalse(((Error)response.getEntity()).getMessage().contains("1"));
```
Comment (rushtong) [inline src/test/java/org/broadinstitute/consent/http/resources/DatasetResourceTest.java:738]: Nice 👍🏽

**Review by rushtong — APPROVED** [review]
Thanks for the changes! 👍🏽

**Review by JVThomas — APPROVED** [review]
Tested locally, batch endpoint is working as intended.
